tracpro/contacts/utils.py

```diff
@@ -0,0 +1,271 @@
+                                                  fields=temba_contact.fields,
+                                                  groups=temba_contact.groups)
+        except TembaBadRequestError as e:
+            temba_err_msg = ''
+            for key, value in e.errors.iteritems():
+                temba_err_msg += key + ': ' + value[0]
+            logger.warning("Unable to create contact %s on RapidPro. Error: %s" % (temba_contact.name, temba_err_msg))
+            return
+
+        # update our contact with the new UUID from RapidPro
+        contact.uuid = temba_contact.uuid
+        contact.save()
+
+    elif change_type == ChangeType.updated:
+        # fetch contact so that we can merge with its URNs, fields and groups
+        remote_contact = client.get_contacts(uuid=contact.uuid)[0]
+        local_contact = contact.as_temba()
+
+        if temba_compare_contacts(remote_contact, local_contact):
+            merged_contact = temba_merge_contacts(
+                local_contact, remote_contact, mutex_group_sets)
+
+            # fetched contacts may have fields with null values but we can't
+            # push these so we remove them
+            merged_contact.fields = {k: v
+                                     for k, v in six.iteritems(merged_contact.fields)
+                                     if v is not None}
+
+            try:
+                client.update_contact(contact=merged_contact.uuid,
+                                      name=merged_contact.name,
+                                      language=merged_contact.language,
+                                      urns=merged_contact.urns,
+                                      fields=merged_contact.fields,
+                                      groups=merged_contact.groups)
+            except TembaBadRequestError as e:
+                temba_err_msg = ''
+                for key, value in e.errors.iteritems():
+                    temba_err_msg += key + ': ' + value[0]
+                logger.warning("Unable to update contact %s on RapidPro. Error: %s" %
+                               (merged_contact.name, temba_err_msg))
+
+    elif change_type == ChangeType.deleted:
+        try:
+            client.delete_contact(contact.uuid)
+        except TembaBadRequestError as e:
+            temba_err_msg = ''
+            for key, value in e.errors.iteritems():
+                temba_err_msg += key + ': ' + value[0]
+            logger.warning("Unable to delete contact %s on RapidPro. Error: %s" % (merged_contact.name, temba_err_msg))
+
+
+def temba_merge_contacts(first, second, mutex_group_sets):
+    """
+    Merges two Temba contacts, with priority given to the first contact
+    """
+    if first.uuid != second.uuid:  # pragma: no cover
+        raise ValueError("Can't merge contacts with different UUIDs")
+
+    # URNs are merged by scheme
+    first_urns_by_scheme = {u[0]: u[1] for u in [urn.split(':', 1) for urn in first.urns]}
```

Comment:
It looks like this will discard all but the last urn for each scheme.  E.g. if there are two urns starting with 'http:', only the last one will end up in first_urns_by_scheme.  Is that a problem?

Same for the next line.